RFCs/0015-Telemetry.md

```diff
@@ -72,8 +74,71 @@ The Microsoft distribution of vstest is licensed with the [Link](https://www.vis
+
+## Telemetry Design
+### Collecting Data
+* Collect Telemetry Data Points in various process(testhost,datacollector etc) and send it to vstest.console process where it will be uploaded or send back to users in design mode. 
+
+**For eg:**
+* In Test Host Process:
+We have to collect how much time does each executor took to run test, Time taken to load adapters etc.
+
+* In Vstest.console:
+We have to collect Total Discovery Time taken, Total Tests Run in case of parallel scenarios.
+
+#### Aggregating data points in Test Host process
+
+* It may happen that TestHost is on a newer version whereas vstest.console is on older version. So, TestHost process should not collect Metrics by default. So, it should only collect Metrics when users give consent to collect Metrics.
+So, for sending consent from Vstest.console process to Test Host, Command line argument i.e. boolean **TelemetryOptedIn** is sent to TestHost Process from Vstest.console process.
+
+The interface [IRequestData](https://github.com/Microsoft/vstest/blob/master/src/Microsoft.TestPlatform.ObjectModel/Client/Interfaces/IRequestData.cs) has been has been exposed to TestHost process which contains [IMetricCollection](https://github.com/Microsoft/vstest/blob/master/src/Microsoft.TestPlatform.ObjectModel/Client/Interfaces/IMetricsCollection.cs) which will collect the Metrics in a dictionary.
 
+#### Sending Metrics from TestHost process to vstest.console process
+Currently, At the end of Discovery Complete, we are sending **TestDiscovery.Complete** message event along with DiscoveryCompletePayload, so we will add our collected Metrics along with [DiscoveryCompletePayload](https://github.com/Microsoft/vstest/blob/master/src/Microsoft.TestPlatform.CommunicationUtilities/Messages/DiscoveryCompletePayload.cs). Similar will be done at end of **TestExecution.Complete** event where we will add the Metrics in [TestRunCompleteEventArgs](https://github.com/Microsoft/vstest/blob/master/src/Microsoft.TestPlatform.ObjectModel/Client/Events/TestRunCompleteEventArgs.cs). This helps us in sending whole metrics in one go and helps us to decrease performance overhead of sending messages from test host process to vstest.console process.
 
+In VsTestConsole process, the metrics which will be received from TestHost process will be aggregated with the VsTestConsole own metrics.
 
+### Publishing Data
 
+For Publishing data, there are two scenarios:
+* Design Mode: We aggregate all the data in vstest.console and send to various IDE's(VS, VSCode etc) giving the IDE's options to add this telemetry data along with thier own Telemetry.
 
+* Non-Design Mode: We will be extending support for it soon.
```

Comment:
@pvlakshm Do you have any timeline in mind for `soon` here? I personally think this is very important because to get bigger picture on usage and we have various client(LUT, TE, Rider, dotnet test, Cloudbuild/CloudTest, etc..).

Reply:
/cc @cltshivash

Reply:
I would not rush this - and instead leave it to the clients to collect, and report, usage of the TestPlatform.